presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/JdkLdapAuthenticatorClient.java

```diff
@@ -0,0 +1,192 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSet.Builder;
+import io.airlift.log.Logger;
+import io.airlift.security.pem.PemReader;
+import io.prestosql.spi.security.AccessDeniedException;
+
+import javax.inject.Inject;
+import javax.naming.AuthenticationException;
+import javax.naming.NamingEnumeration;
+import javax.naming.NamingException;
+import javax.naming.directory.DirContext;
+import javax.naming.directory.SearchControls;
+import javax.naming.directory.SearchResult;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509TrustManager;
+
+import java.io.File;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static io.prestosql.plugin.password.jndi.JndiUtils.createDirContext;
+import static java.util.Objects.requireNonNull;
+import static javax.naming.Context.INITIAL_CONTEXT_FACTORY;
+import static javax.naming.Context.PROVIDER_URL;
+import static javax.naming.Context.REFERRAL;
+import static javax.naming.Context.SECURITY_AUTHENTICATION;
+import static javax.naming.Context.SECURITY_CREDENTIALS;
+import static javax.naming.Context.SECURITY_PRINCIPAL;
+
+public class JdkLdapAuthenticatorClient
+        implements LdapAuthenticatorClient
+{
+    private static final Logger log = Logger.get(JdkLdapAuthenticatorClient.class);
+
+    private final Map<String, String> basicEnvironment;
+    private final Optional<SSLContext> sslContext;
+
+    @Inject
+    public JdkLdapAuthenticatorClient(LdapConfig ldapConfig)
+    {
+        String ldapUrl = requireNonNull(ldapConfig.getLdapUrl(), "ldapUrl is null");
+        if (ldapUrl.startsWith("ldap://")) {
+            log.warn("Passwords will be sent in the clear to the LDAP server. Please consider using SSL to connect.");
+        }
+
+        this.basicEnvironment = ImmutableMap.<String, String>builder()
+                .put(INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory")
+                .put(PROVIDER_URL, ldapUrl)
+                .put(REFERRAL, ldapConfig.isIgnoreReferrals() ? "ignore" : "follow")
+                .build();
+
+        this.sslContext = Optional.ofNullable(ldapConfig.getTrustCertificate())
+                .map(JdkLdapAuthenticatorClient::createSslContext);
+    }
+
+    @Override
+    public void validatePassword(String userDistinguishedName, String password)
+            throws NamingException
+    {
+        createUserDirContext(userDistinguishedName, password).close();
+    }
+
+    @Override
+    public boolean isGroupMember(String searchBase, String groupSearch, String contextUserDistinguishedName, String contextPassword)
+            throws NamingException
+    {
+        DirContext context = createUserDirContext(contextUserDistinguishedName, contextPassword);
```

Comment:
try with resources?

If `DirContext` is not closeable then:

```
try (CloseableWrapper<DirContext> context = wrap(createUserDirContext(contextUserDistinguishedName, contextPassword)) {
```
?

Reply:
I didn't know about `CloseableWrapper` I'll give it a try

Reply:
I just make up this name, you would need to write your own. My point was wrap non closeable object as closeable so you could use try-with-resources.

Reply:
This entire code base is already using try/finally, so this follows the existing code.